File: apps/go/chapter02/arrays_test.go
package chapter02_test

import (
	"testing"

	"github.com/k2works/getting-started-algorithm/go/chapter02"
)

func TestMaxOf(t *testing.T) {
	got := chapter02.MaxOf([]int{172, 153, 192, 140, 165})
	if got != 192 {
		t.Errorf("MaxOf(...) = %d; want 192", got)
	}
}

func TestReverseArray(t *testing.T) {
	a := []int{2, 5, 1, 3, 9, 6, 7}
	chapter02.ReverseArray(a)
	expected := []int{7, 6, 9, 3, 1, 5, 2}
	for i, v := range a {
		if v != expected[i] {
			t.Errorf("ReverseArray index %d = %d; want %d", i, v, expected[i])
		}
	}
}

func TestCardConv(t *testing.T) {
	tests := []struct {
		x, r     int
		expected string
	}{
		{29, 2, "11101"},
		{255, 16, "FF"},
	}
	for _, tt := range tests {
		got := chapter02.CardConv(tt.x, tt.r)
		if got != tt.expected {
			t.Errorf("CardConv(%d,%d) = %q; want %q", tt.x, tt.r, got, tt.expected)
		}
	}
}

func TestPrime1(t *testing.T) {
	got := chapter02.Prime1(1000)
	if got != 78022 {
		t.Errorf("Prime1(1000) = %d; want 78022", got)
	}
}

func TestPrime2(t *testing.T) {
	got := chapter02.Prime2(1000)
	if got != 14622 {
		t.Errorf("Prime2(1000) = %d; want 14622", got)
	}
}

func TestPrime3(t *testing.T) {
	got := chapter02.Prime3(1000)
	if got != 3774 {
		t.Errorf("Prime3(1000) = %d; want 3774", got)
	}
}

File: apps/c/chapter02/arrays_test.c
#include <stdio.h>
#include <string.h>
#include "../include/test_helper.h"
#include "arrays.h"

void test_max_of(void) {
    int a[] = {172, 153, 192, 140, 165};
    ASSERT_EQ_INT(192, max_of(a, 5));
}

void test_reverse_array(void) {
    int a[] = {2, 5, 1, 3, 9, 6, 7};
    reverse_array(a, 7);
    int expected[] = {7, 6, 9, 3, 1, 5, 2};
    for (int i = 0; i < 7; i++) {
        ASSERT_EQ_INT(expected[i], a[i]);
    }
}

void test_card_conv(void) {
    char buf[64];
    card_conv(29, 2, buf, sizeof(buf));
    ASSERT_EQ_STR("11101", buf);
    card_conv(255, 16, buf, sizeof(buf));
    ASSERT_EQ_STR("FF", buf);
}

void test_prime1(void) {
    ASSERT_EQ_INT(78022, prime1(1000));
}

void test_prime2(void) {
    ASSERT_EQ_INT(14622, prime2(1000));
}

void test_prime3(void) {
    ASSERT_EQ_INT(3774, prime3(1000));
}

int main(void) {
    test_max_of();
    test_reverse_array();
    test_card_conv();
    test_prime1();
    test_prime2();
    test_prime3();
    TEST_SUMMARY();
    return 0;
}
